### PR TITLE
StratFeeManagerInitializable, VaultFactory, update VaultV7

### PR DESCRIPTION
**Pull request by roman-monk**
Fixed init in VaultV7.
StratFeeManager with init function instead of constructor.
StrategyConvex as a sample.

address prediction is not needed, vault deployment would be something like:
```
let tx;
const Factory = await ethers.getContractAt("BeefyVaultV7Factory", factoryAddress);
tx = await Factory.cloneVault();
await tx.wait();
const vaultAddress = await getProxyCreatedAddress(Factory, tx);

tx = await Factory.cloneContract(strategyImplAddress);
await tx.wait();
const stratAddress = await getProxyCreatedAddress(Factory, tx);

const Vault = await ethers.getContractAt(config.vaultName, vaultAddress);
tx = await Vault.initialize(stratAddress, mooName, mooSymbol, delay);
await tx.wait();

const Strategy = await ethers.getContractAt(strategyName, stratAddress);
tx = await Strategy.initialize(...stratArgs);
await tx.wait();
```

### CODE REVIEW

File: contracts/BIFI/vaults/BeefyVaultV7.sol [hunk 1]
```diff
@@ -31,12 +31,6 @@ contract BeefyVaultV7 is ERC20Upgradeable, OwnableUpgradeable, ReentrancyGuardUp
 
     event NewStratCandidate(address implementation);
     event UpgradeStrat(address implementation);
-    event BeefyVaultV7Initialized(
-        IStrategyV7 strategy,
-        string name,
-        string symbol,
-        uint256 approvalDelay
-    );
```
Comment (MirthFutures) [inline contracts/BIFI/vaults/BeefyVaultV7.sol:34]: Should we add Deposit and Withdraw events?

File: contracts/BIFI/strategies/Curve/StrategyConvex.sol [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+    address public native;
+    address public depositToken;
+
+    // Third party contracts
+    address public gaugeFactory;
+    address public rewardsGauge;
+    address public pool;
+    uint public poolSize;
+    uint public depositIndex;
+    bool public useUnderlying;
```
Comment (MirthFutures) [inline contracts/BIFI/strategies/Curve/StrategyConvex.sol:30]: If you move the bools ahead of the uint you actually save a bit of gas since it will pack the bools in the same slot as pool

File: contracts/BIFI/strategies/Curve/StrategyConvex.sol [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+
+    // if no CRV rewards yet, can enable later with custom router
+    bool public crvEnabled;
+    address public crvRouter;
+
+    // if depositToken should be sent as unwrapped native
+    bool public depositNative;
+
+    bool public harvestOnDeposit;
+    uint256 public lastHarvest;
```
Comment (MirthFutures) [inline contracts/BIFI/strategies/Curve/StrategyConvex.sol:53]: Should we change all lastHarvest to uint32? uint32 max number is sometime in 2108.

**Review by sircryptomaxi — COMMENTED** [review]
Wdyt about extending the vault with ERC4626 so we're compliant while keeping all our infrastructre still?

File: contracts/BIFI/strategies/Curve/StrategyConvex.sol [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+import "@openzeppelin-4/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import "../../interfaces/common/IUniswapRouterETH.sol";
+import "../../interfaces/common/IWrappedNative.sol";
+import "../../interfaces/curve/ICurveSwap.sol";
+import "../../interfaces/curve/IGaugeFactory.sol";
+import "../../interfaces/curve/IRewardsGauge.sol";
+import "../Common/StratFeeManagerInitializable.sol";
+
+contract StrategyConvex is StratFeeManagerInitializable {
```
[the remark this answers is not in the document]
Comment (roman-monk) [reply contracts/BIFI/strategies/Curve/StrategyConvex.sol:15]: there is a very slight chance it will be needed, practically we never use it outside of bsc, its not even deployed on other chains

**Comment by roman-monk** [comment]
> Wdyt about extending the vault with ERC4626 so we're compliant while keeping all our infrastructre still?

yeah very smart to make it as a wrapper

**Comment by roman-monk** [comment]
@sircryptomaxi @MirthFutures 
about gas saving, its more complex than just packing or ++i to save some real amounts.

changing `lastHarvest` type could actually cost more because of the type conversion. 
but the simple update of lastHarvest on every harvest() costs 5000 gas = 0.0001 ether on 20 gwei.
thats like $0.2 on every single harvest.
just to compare with 5-10 gas saved on ++i, or some weird `unchecked {i++}` (which is easy to miss and difficult to read) on our 3-iteration-max loops.

`feeConfig.getFees()` costs 15k-30k gas, i checked few on tenderly on different chains, these values practically "never" change but we read it on every harvest and thats $1 every time.
could this be cached? or can we drop "string label" which is not used on chain but probably breaks etherscan and costs $$ for reading and sending (not entirely sure).

Deposit and Harvest events make 4-5 external calls for balanceOf though some values not changed in between.

HOD actually deposits twice which adds insane gas, because most of the gas is spent on external deposits and swaps, almost unoptimizable. 

or 3 transfers on harvest, could be consolidated in strat and send later as a batch? maybe without callFee and it could be hardcoded, but strategist and beefyFee could be calculated and sent later saving also on getFees().

some group of strats like Aura could send BAL and AURA rewards to the 1 swapper and get WETH back later when swapper is harvested. that would save lots of gas on swaps but adds more complexity.
1 swap is at least 100,000 gas (0.002 eth on 20 gwei). 
2 swaps for both tokens, 1 swap from native, then addLiquidity, thats close to 0.01 eth total on every harvest of every strat.

so in short amount of external calls matters, and optimization would require quite refactoring
